### ISSUE TITLE
Build fails

### ISSUE THREAD

**Issue by probonopd** [giulioz/mini-jv880#1]
Hi @giulioz, kudos for this project!
Must have been quite a challenge to transplant another synth engine into it.

Currently the build fails; possibly due to a missing file, `usbkompletekontrol.h`?

```
  CPP   properties.o
  CPP   propertiesfile.o
  CPP   propertiesfatfsfile.o
  CPP   propertiesbasefile.o
  AR    libproperties.a
+ cd ../../../..
+ cd ..
+ cd src
+ make clean
  CLEAN  /home/runner/work/mini-jv880/mini-jv880/src
+ make -j
make: *** No rule to make target 'circle/usb/usbkompletekontrol.h', needed by 'main.o'.  Stop.
Error: Process completed with exit code 2.
```

**Comment by giulioz** [comment]
Hey thanks a lot! It was actually quite straightforward thanks for your project :)
Yeah the file is missing since I had to put that file inside the Circle folder, which is in the gitignore. I can make the file public but it's tied to the Native Instrument Komplete Kontrol keyboard, the one I showed in my video (https://www.youtube.com/watch?v=7LWaYHTNIDI). I can make a branch that doesn't require it...

**Comment by probonopd** [comment]
Yes, would appreciate if you could make it optional. 

What Native Instrument Komplete Kontrol specific functionality did you implement? Would it be beneficial for MiniDexed too? Unfortunately, I don't have such a controller but I hear good things about S series MK3 (poly aftertouch; still to be implemented in MiniDexed).

**Comment by plamikcho** [comment]
Hi @giulioz, congrats for your project! I am trying to create a version compatible with the MiniDexed hardware. My setup is RPi4, SSD1306, PCM5102. I put some of the MiniDexed code for the UI and the buttons but I have troubles with the LCD - I cannot get it working on the LCD. I do not fully understand the lcdbuffer and the way it works in Nuked-SC55. MiniDexed LCD uses characters but SC-55's are pixels. Currently I can see the JV880 screen emulation on the HDMI but it is empty - no characters are printed. 

```
for (size_t y = 0; y < lcd_height; y++) {
		for (size_t x = 0; x < lcd_width; x++) {
			m_pMiniJV880->screenUnbuffered->SetPixel(x + 800, y + 100, lcd_buffer[y * lcd_width + x]);
		}
	}
```

Also I do not have the "jv880_nvram.bin" file - could this be the issue? The code is here: https://github.com/plamikcho/mini-jv880/tree/codespace-humble-goldfish-r79xv4x557hp4jx

**Comment by giulioz** [comment]
Hi @plamikcho, thank you!

The SSD1306 should be very similar to the LCD of the Komplete Kontrol controller I'm using. What I did is "resizing" the pixels output of the emulator to fit into the smaller screen (https://github.com/giulioz/mini-jv880/blob/main/src/minijv880.cpp#L244). You can modify that code to send it to your screen instead. Keep in mind that my screen had a weird protocol with one byte for each column of the screen, so maybe you need to change the algorithm there.

You should not need the `jv880_nvram.bin` file, it's just a copy of the content of the NVRAM after a factory reset to have it ready on start. I've uploaded it here anyway: https://drive.google.com/file/d/18kvHI3vs2dEgYxPViERpdJWWFUvGpX0F/view?usp=sharing

I've looked at your code, great work making it more generic! Feel free to submit a PR if you want :) I'm unfortunately working on some other emulations right now so I don't have time yet...

Are you saying that you see no characters in the HDMI display as in it stays completely green? It might be that either the emulator is actually not running, some error occurred or the screen is not updated correctly. Maybe you can put some logging points in some locations to see what's happening. Do you have the correct roms?

**Comment by plamikcho** [comment]
Thank you for the fast response! Unfortunately C++ is not my technology stack and probably that's the main reason to not be able to find the issue. I cannot see errors in the log messages, the value returned from mcu.startSC55 is 0. I am using the correct ROMs - I have Nuked SC-55 running on macos. I guess the screen is not updated properly - but it was like that before my modifications, at first I commented the komplete kontrol references to unblock the build and it was green without any characters rendered. Here are some pictures:
<img width="1101" alt="Screenshot 2024-10-18 at 13 56 37" src="https://github.com/user-attachments/assets/59b6e189-58ef-4693-8a7c-9733bc64c2c7">
![IMG_20241018_151713](https://github.com/user-attachments/assets/72275070-a12a-40e2-9978-4079c1ab86b4)
![IMG_20241018_151716](https://github.com/user-attachments/assets/3cc070e1-b894-4aab-b2b9-db599090cb33)

UPDATE: I printed the emulator LCD_Data with the characters using the same code:
```
                 for (int i = 0; i < 2; i++)
		{
			std::string str = "MCU LCD_Data Character row " + std::to_string(i);
			for (int j = 0; j < 24; j++)
			{
				uint8_t ch = m_pMiniJV880->mcu.lcd.LCD_Data[i * 40 + j];
				str.append(" ");
				str.append(std::to_string(ch));
			}
			LOGNOTE(str.c_str());
		}
```
All cells contained the same value of 32 and this never changes. I guess the emulator is not running by some reason...

**Comment by giulioz** [comment]
You can do the following things:
- Show me the content of your SD card
- Put a log point after https://github.com/plamikcho/mini-jv880/blob/codespace-humble-goldfish-r79xv4x557hp4jx/src/minijv880.cpp#L158 ("initialised")
- Put a log point after https://github.com/plamikcho/mini-jv880/blob/codespace-humble-goldfish-r79xv4x557hp4jx/src/minijv880.cpp#L219 ("has frames to render")
- Put a log point after https://github.com/plamikcho/mini-jv880/blob/codespace-humble-goldfish-r79xv4x557hp4jx/src/minijv880.cpp#L237 ("emu cycle")

**Comment by plamikcho** [comment]
SD card content:
<img width="845" alt="Screenshot 2024-10-18 at 18 01 08" src="https://github.com/user-attachments/assets/9a41b896-b390-4736-af6f-5335c3ffe2e3">
Loggers:
```
nSamples = (int)nFrames * 2;
LOGNOTE("has frames to render: %d", nSamples);
```
```
mcu.MCU_UpdateAnalog(mcu.mcu.cycles);
LOGNOTE("emu cycle: %d", mcu.mcu.cycles);
```
![IMG_20241018_194416](https://github.com/user-attachments/assets/86b90e1e-8213-4662-aedc-642c44a9ff2f)
Loggers are called every 2 seconds to prevent flooding - without the throttling I see rapidly increasing numbers for both.

**Comment by giulioz** [comment]
Ah ok so the emulator is actually running! But all the LCD is full of spaces for some reason…
I have honestly no idea why that is happening but I'm very curious why it doesn't work :(
It's possible that something is going wrong inside the emulator code, there is an "error logging handler" here https://github.com/plamikcho/mini-jv880/blob/codespace-humble-goldfish-r79xv4x557hp4jx/src/emulator/mcu.cpp#L43 but it uses printf so I doubt it gets outputed in the logs, maybe you need to convert that into a LOGNOTE

**Comment by plamikcho** [comment]
The error trap appears only once after initialisation with message "trap 3 18fc". No other error messages.

**Comment by giulioz** [comment]
It's crashing in this instruction here in the emulator, unfortunately I have no idea why :(

<img width="459" alt="image" src="https://github.com/user-attachments/assets/98050fde-09bf-4347-9d06-e1f48635b1ac">

It could be because of the multicore adaptation I did that it doesn't like the RPI4 for some reason, or some uninitialized variable that for some reason works on my machine instead. Or maybe the rom is not the correct one?

**Comment by plamikcho** [comment]
You were right about the rom - rom2 was wrong. This was the only thing I haven't tried :) There are 2 versions of this rom 1.00 and 1.01, I used the 2nd one and it didn't work. Tried with the 1.0.0 and the emulator started normally. Somehow the 1.0.1 works in the desktop version of the emulator? Now I have to find a way to print the emulator LCD on the SSD1306, I think that using LCD_Data and void CUserInterface::LCDWrite (const char *pString) should be easier than scaling the pixels.
![IMG_20241019_192139](https://github.com/user-attachments/assets/07e1f94d-1a02-4e56-892b-517c9b9526f9)
<img width="663" alt="Screenshot 2024-10-19 at 19 24 38" src="https://github.com/user-attachments/assets/2035eae0-5b89-459e-b497-3582f5a9eace">

**Comment by giulioz** [comment]
Yeah that's because I crippled the emulator for this version a lot to make it faster, so probably I tuned it too much to work on the 1.0.0 only.

> I think that using LCD_Data and void CUserInterface::LCDWrite (const char *pString) should be easier than scaling the pixels

That could work! Keep in mind thought that some characters maybe are using some custom fonts, so you would have to implement that as well.

Does audio/midi work for you already?

**Comment by plamikcho** [comment]
I haven't tested with my keyboard yet, probably will try tomorrow. I am able to print on the LCD but the rows are not right. My LCD seem to support only 20 columns but minijv880 has 24? Currently this is the best I can print :)
```
      CString Msg ("\x1B[H\E[?25l");
	for (int i = 0; i < 2; i++)
	{
		for (int j = 0; j < 24; j++)
		{
			uint8_t ch = m_pMiniJV880->mcu.lcd.LCD_Data[i * 40 + j];
			std::string str(1, ch);
			const char *pString = str.c_str();
			Msg.Append(pString);
		}
	}
	LCDWrite(Msg);
```
![IMG_20241019_212233](https://github.com/user-attachments/assets/ac1fe08a-9494-443c-b9bb-e80d8db7e053)

**Comment by giulioz** [comment]
Yeah you probably need to output pixels and horizontally compress the resolution like I did on my code :(

**Comment by plamikcho** [comment]
Just tested minijv880 with my keyboard. On RPi4 the sound is terrible, full of glitches and distorted. Also there is some background noise. On RPi3 the sound is much better but still on most of the patches there are distortions and glitches. Here are some recordings:
https://audio.com/plamen-dzorev/audio/jv880-test-dowah
https://audio.com/plamen-dzorev/audio/jv880-test-trumpet
https://audio.com/plamen-dzorev/audio/jv880-test-jazzorgan
These are some of the playable ones. This is how it looks in Audacity - there are interruptions?
<img width="1413" alt="Screenshot 2024-10-20 at 10 48 40" src="https://github.com/user-attachments/assets/acd5db80-151c-4b23-81e7-02cd10e2ba98">

**Comment by giulioz** [comment]
Yeah that's unfortunately the difficult part…
The emulator is quite slow and I did already a lot to optimize it. The emulator has to generate all the samples in the buffer before the playback of the previous buffer finishes (which is ~10ms). If it can't keep up you will hear audible click glitches.
In my PI Zero 2 it was not fast enough, so what I did was split the cpu and sound chip emulator in two cores. This kinda works but it also introduces some other strange race conditions. It's possible that it doesn't work properly in the RPi4, but it's also possible that the RPi4 is actually fast enough to run it on a single core.

You could try it by changing the loop in this line (https://github.com/plamikcho/mini-jv880/blob/codespace-humble-goldfish-r79xv4x557hp4jx/src/minijv880.cpp#L241) by uncommenting that PCM_Update and commenting the one in the other core

**Comment by plamikcho** [comment]
This worked for RPi4 - most of the patches are playable. Unfortunately after a few minutes of playing the Pi gets hot and probably limits its CPU to protect from overheat which results again in glitches - more notes, more glitches. RPi3 cannot process all the stuff on a single core and the sound has glitches all the time. To get smooth sound I removed the LCD output to the HDMI and left only the SSD1306 code. Maybe the solution is to split on separate cores with some synchronisation to prevent the racing conditions?
Otherwise the JV880 is excellent :)
https://audio.com/plamen-dzorev/audio/jv880test-turboa

**Comment by plamikcho** [comment]
No luck with the glitches till now, the only thing I noticed is that once the sound device and MCU get out of sync, it stays like that and it cannot get back in sync, usually I am changing the patch with a lightweight one (which does not cause glitches like bass or a guitar), playing some time single notes only and eventually the glitches may disappear when switch back to the previous patch. The most calculation intensive patches are pianos, fantasia, organs. Better performing ones are guitars, bass, sax. Next thing I will try is to add a cooling fan - maybe this will keep the Pi in the optimal performance zone (first 5-10 minutes of playing are without glitches, the problems begins after the Pi gets hot). Aaaand ... another idea to solve the issue is to buy RPi5 which is at least 2x faster than RPi4 - probably the emulator will run without glitches :)
![IMG_20241025_143624](https://github.com/user-attachments/assets/5e6d2518-c8f8-410f-990a-374868cf6126)

**Comment by giulioz** [comment]
@plamikcho cool buttons!
Are you running the PCM and CPU emulators in a single core sequentially now?
One thing that could help you is having the CPU run at full speed. IIRC the Circle library underclocks the Pi CPU by default.

**Comment by plamikcho** [comment]
Yes, I am running on a single core on RPi4. I will try at full speed but after I get a cooling fan, otherwise it will overheat. It runs fine until it gets hot so I guess the speed is enough but with the CPU temp rising it reduces the speed and the glitches happen. I saw that you put spin locks inside PCM - can I remove them when working in single core mode?

**Comment by giulioz** [comment]
Yeah you can remove them

**Comment by plamikcho** [comment]
Slowly going into the right direction :) After removing the spinlocks I have between 10 and 15 mins of good sound without glitches on all patches (tried pianos and organs which are "heavier"). Checked the RPi speed - it is already set to maximum
```
CKernel::CKernel (void)
:	CStdlibAppStdio ("minijv880"),
	m_Config (&mFileSystem),
	m_GPIOManager (&mInterrupt),
 	m_I2CMaster (CMachineInfo::Get ()->GetDevice (DeviceI2CMaster), TRUE),
	m_pSPIMaster (nullptr),
	m_pJV880 (0),
	m_CPUThrottle (CPUSpeedMaximum)
{
	s_pThis = this;

	// mActLED.Blink (5);	// show we are alive
}
```
Ordered a 30mm cooling fan for RPi. Also noticed that the Pi consumes a lot of power - probably 1A, it drains my external power bank really fast.

**Comment by giulioz** [comment]
The 10-15 mins really makes me think it's just throttling. But nice!!
Do you also have a metal heatsink?

**Comment by plamikcho** [comment]
Yes, I have heatsinks on all the 3 chips - that's why it plays for 10 mins. It should be throttling, because the whole housing gets really hot - and it's plastic! Also it happens to restart sometimes - probably of overheating.

**Comment by michelemaroni** [comment]
Hi all,

stumbled across this project after finding the youtube video while I was searching for Nuked-SC55 from [Zynthian](https://discourse.zynthian.org/t/nuked-sc55-and-kontakt-samples/10203).
I'm really thrilled by this project and I hope I'll manage to test it soon in one of my RPis.

For this reason I'd like to offer to contribute.
I've got no expertise in C/C++ programming, but I'm a python developer.
I regulary use Linux, and Bash, and I can follow recipes for compiling stuff if the instructions are provided.
I could contribute to the build instructions, and probaly testing builds for RPI4 and 5.

Reading through this post and looking at the reported performance issues,
the most concrete contribution that i feel i can make at the moment is bringing to your attention the youtube channel of the [Audio Developers Conference](https://www.youtube.com/@audiodevcon), and also in general looking for the talks made by Timur Doumler in [this](https://www.youtube.com/watch?v=vn7563IAQ_E) and other C/C++/AudioDev conferences. 

Hope this will help, and I look forward to hearing what I can do to help in getting builds and realeases available for the community.

Thanks @giulioz for setting up this project, @probonopd for the terrific work on minidexed (I'll also test that soon), and @plamikcho for the reports and  building efforts!

**Comment by probonopd** [comment]
Can it be built and run on (which?) RPi models with standard USB class compliant MIDI controllers nowadays?

**Comment by plamikcho** [comment]
This fork https://github.com/plamikcho/mini-jv880 can be built and works on this hardware: RPi 4, DAC (PCM5102), LCD SSD1306 with a standard USB MIDI controller (mine is Arturia KeyStep 37). The Pi needs to be cooled - metal heatsink + fan, otherwise it gets throttled after 10 mins and the sound is with glitches. Tested with the setup below:
![IMG_20241107_140134](https://github.com/user-attachments/assets/687d77f6-28aa-4259-ad5b-e6a11b0499e8)
You need a good power supply (power bank or adapter, at least 2A). I don't think it will work on something below the RPi4 without further optimization.
**EDIT:** Forgot to mention that even without the buttons if you have a working MiniDexed with rotary encoder on its default pins with the above hardware (RPi4, SSD1306) you can play :) Patches are changed with the rotary encoder, by default it is set to midi channel 1.

**Comment by michelemaroni** [comment]
Thanks for the updates @plamikcho 

I cloned your repo, trying to figure out how to build, but i have three questions though:

- Does the SSD1306 128x64 i2c OLED works with this build? Does the screen resolution need to be changed in the code?
- Can it be built on x86_64 or for the moment cross compilation is not set up in the build scripts?
- To build a full image to be dumped on an SD, the script to be run is build_sd.sh, with the two block in lines 12-19 and 21-29 uncommented, right?

BTW, the keyboard you set up is very cool, do you have instructions on how to build that as well?

Thanks!

**Comment by plamikcho** [comment]
It works with SSD1306 compatible display only - it needs 20 columns to display properly. I have used a simple workaround to write 24 columns in 20 - skipping 4 columns, optimized for the Patch screen (missed the brackets, spaces and sometimes a letter from the patch name). It is hardcoded - configuration is not taken into account.
I am building it for Aarch64 only, you can have a build for others too.
I followed the build instructions for MiniDexed here: https://github.com/probonopd/MiniDexed/wiki/Development (the first one in the cloud using codespaces). 
```
git submodule update --init --recursive
sh -ex submod.sh # Get particular git commits of the submodules
wget -q https://developer.arm.com/-/media/Files/downloads/gnu-a/10.3-2021.07/binrel/gcc-arm-10.3-2021.07-x86_64-aarch64-none-elf.tar.xz
tar xf *-aarch64-none-elf.tar.xz
# wget -q https://developer.arm.com/-/media/Files/downloads/gnu-a/10.3-2021.07/binrel/gcc-arm-10.3-2021.07-x86_64-arm-none-eabi.tar.xz
# tar xf *-arm-none-eabi.tar.xz
export PATH=$(readlink -f ./gcc-*aarch64-none*/bin/):$PATH
RPI=4 bash -ex build.sh
```
I will post more pictures of the keyboard + some instructions how to build it.
Also you will need the jv880_nvram.bin file - it is not included with the ROMs, you can download it from the Google drive link shared in a previous comment.

**Comment by plamikcho** [comment]
The keyboard is built with 9 buttons - ordered from AliExpress. You can search for "12x12x10 Tact Switch with plastic cap". It should look like this:
<img width="1129" alt="Screenshot 2024-11-10 at 20 04 41" src="https://github.com/user-attachments/assets/ecbfac07-88f6-48e7-9fa5-1924da697667">
Next thing you need is a universal PCB - at least 60x60 mm, I don't have the size because I've cut mine to fit 3x3 button grid.
![double-sided-prototype-printed-circuit-board-5x7cm](https://github.com/user-attachments/assets/52e7febb-703e-4453-afbf-d5d0ec4151e8)
After you solder the buttons, choose the pins to be grounded - the top or the bottom ones, doesn't matter, just make sure to connect the same pins on all buttons to the ground. I used simple bridges on the PCB. For the input pins, solder an individual wire to each button. Then select some free GPIOs - I've selected the most convenient for me, next to each other, 3 groups and connected with Dupont connectors. 
![IMG_20241024_134738](https://github.com/user-attachments/assets/6a0b03f6-36dc-4e78-8033-4c546f1efa0e)
The last thing to do is to map the buttons to a particular function on the JV-880 - it has 14 buttons total, 1 is used on the rotary encoder (Enter), and 9 on the PCB and I just skipped 3 buttons - Mute, Monitor and Compare. It was difficult for me to follow which button on which GPIO to connect, so I connected them sequentially and then used a debug print to see what code is being sent when I press a button.
For the 9 buttons I used the codes are:

10 DATA                         | 100 UTIL          | 1000 SYSTEM
1 LEFT                            | 200 PREVIEW | 2 RIGHT
400 PATCH PERFORM | 800 EDIT         | 4 TONE SELECT
I have filled the minijv880.ini with my configuration - you can see which GPIOs are used.
![IMG_20241025_103038](https://github.com/user-attachments/assets/8c55b933-af47-43e1-9087-290f7bf13daf)
Unfortunately, after mounting of the fan the case cannot be closed and I had to use a tape to fill the gap of 6-7 mm. If you decide to buy a similar case, maybe choose one with height at least 42 mm to be able to fit with the rotary encoder, the LCD and the fan.

**Comment by blixabloxa** [comment]
Great stuff Plamikcho!

**Comment by michelemaroni** [comment]
Hi @plamikcho,

I am attempting the build from [this branch of your fork](https://github.com/plamikcho/mini-jv880/tree/codespace-humble-goldfish-r79xv4x557hp4jx).

Following the instruction you provided, the build.sh script works.
However, the full (so with  lines 12-19 uncommented , except the real comments) build_sd.sh does not:
it fails on line 24 because it misses './src/performance.ini'.

`+ cp ./src/config.txt ./src/minijv880.ini ./src/kernel8-rpi4.img ./src/performance.ini sdcard/
cp: cannot stat './src/performance.ini': No such file or directory`

Could you please add it to your fork?

Thanks!

**Comment by plamikcho** [comment]
You want to use the build_sd.sh script? It contains lines from the MiniDexed dev guide here: https://github.com/probonopd/MiniDexed/wiki/Development
The lines are commented because you don't need to execute this on every build. You can just remove performance.ini, this is related to MiniDexed (the code was copied from MiniDexed and there are leftovers). I removed this file from build_sd.sh (but left the commented areas).
Here is a working build (the one I am using). It works only on RPi4/SSD1306/rotary encoder. You have to add the ROMs yourself - it is not permitted to distribute them. Use at your own risk!
[minijv880_build.zip](https://github.com/user-attachments/files/17750583/minijv880_build.zip)
Here is a build with scrolling LCD taking into account the LCD col setting. It should work with 16x2 too (not tested).
[minijv880-20241116.zip](https://github.com/user-attachments/files/17784571/minijv880-20241116.zip)

**Comment by michelemaroni** [comment]
Hi all, sorry for not having replied earlier.

I figure out how to build the SD image. [Minidexed wiki - Building locally](https://github.com/probonopd/MiniDexed/wiki/Development#building-locally) and [Nektos Act](https://github.com/nektos/act) proved very useful for the purpose.

I'm attaching my build of @plamikcho 's fork main branch, which run on RPI 3A (pic below), RPI4 and should run also on RPI 5 (untested).

[MiniJV8802025-03-07.zip](https://github.com/user-attachments/files/19229468/MiniJV8802025-03-07.zip)

![Image](https://github.com/user-attachments/assets/33d450b2-1391-42e0-a538-a4f48a1bd055)

I also noticed that attaching a DSI screen makes the sound sort of "interrupt" when playing multiple notes, and warning/error message appears on the screen.

![Image](https://github.com/user-attachments/assets/4b14f6a2-1046-4577-9a4d-fc9e933489d2).

Overall, I'm very enthusisatic and I hope to see this pproject grow further.

**Comment by plamikcho** [comment]
How is the sound on RPi3? I remember that the single core version worked on RPi4 only, on 3 the sound was with glitches and interruptions.

**Comment by michelemaroni** [comment]
I made mostly a smoke test, so I didn't paid that much attention to that.
However, the integrated heaphone sounded usable, but i experienced audio artifacts with both RPI 3 (integrated headphone)
and RPI4 ([This PCM5122 DAC](https://wiki.geekworm.com/DACPi)) whenever I had a dsi screen attached.

I'm waiting for some DACs to be delivered and I'll come back with some deeper testing.

I would suggest to change also the cpu temperature monitoring behaviour:
check the temperature every 15-30 seconds, and printing warnings only if the cpu temperature exceeds 55°C should provide adequate monitoring while sparing also some resources.

**Comment by probonopd** [comment]
Hello @michelemaroni. I think you are the first person to ever try a DSI screen with MiniDexed. It is not something we have ever considered or tested. Please be aware that the large displays (HDMI or in your case DSI) are only for developing/debugging and are not needed for just playing/configuring the instrument.

**Comment by michelemaroni** [comment]
Hi @probonopd , thanks for clarifying.
Indeed it's just the thing i noticed at my first attempato. DSI/HDMI display are definite an overkill for minidexed and minijv880 projects, and i didn't plan to stably use one. It is still useful to report this behaviour in git issues though.
So anyone willing to test know that dsi/hdmi are not full usable for playing (and probably shouldn't).

**Comment by ichingxs4all** [comment]
![Image](https://github.com/user-attachments/assets/339934b3-212b-4479-b1c1-6ac01a1aef6f)

I am also trying to get the mini-JV880 running on the euroack miniDexed design of Kevin of diyelectromusic.com.
So far I already converted the mt32pi project to run on this design.

[https://diyelectromusic.com/2025/04/07/mt32-pi-on-my-eurorack-minidexed-pcb/](url)

I used the build from @michelemaroni and I can confirm it is running and producing sound ( with glitches ) by using an Arturia Keystep on the usb port of the Zero 2 W

I wonder if I can be of any help to try to optimise the project so it runs on a Zero 2 pi without the audio glitches and using the hardware serial port for midi input instead of the usb port. My goal is to have a euro rack version of the mini-JV880.

**Comment by ichingxs4all** [comment]
Maybe the glitches are caused by these 'unknown write ' errors ?
What does these errors mean ?

**Comment by ichingxs4all** [comment]
![Image](https://github.com/user-attachments/assets/4d27361d-7d3d-4289-9e9f-30744a7b2527)

**Comment by giulioz** [comment]
Hi @ichingxs4all, your project looks very cool!
The unknown write errors are expected: there is an harmless bug somewhere in the JV880 rom in which it writes into the ROM. It has no downside and I verified it happening also in the real hardware.

Regarding the glitches, they happens every time the two threads go out of sync (if you are using the multithreaded version) or when the audio generation becomes slower than the audio output buffer fill rate. I wanted to get into some optimisations (I believe many are possible) but I didn't have time yet. I've resorted to running it in two threads in my version (cpu emulator on one and audio generation on the other), but syncronizing the two causes issues.

**Comment by ichingxs4all** [comment]
Thnx for your response.

I hooked up a RPi4 with a i2s PCM5102 dac and it doesn't have the audio glitches. So now have a reference how it should sound like :-)

I have to dive more deep down into code first to get a better understanding how the midi from the usb device is handed over to the emulator. And then I want to try to implement the hardware / gpio serial midi. I want to make use of the minidexed code for this. 

But I am curious. You were able to have a Zero 2 W pi running it without the glitches. I think the divergence is that your initial setup was using the builtin pwm headphone output and a Native Instrument Control keyboard with its own 'driver' not the usb standard midi driver. Was it ?

**Comment by giulioz** [comment]
> I have to dive more deep down into code first to get a better understanding how the midi from the usb device is handed over to the emulator

The core part is the `postMidiSC55` function, which can send an array of midi bytes to the emulator.
 
> But I am curious. You were able to have a Zero 2 W pi running it without the glitches. I think the divergence is that your initial setup was using the builtin pwm headphone output and a Native Instrument Control keyboard with its own 'driver' not the usb standard midi driver. Was it ?

Well.. I had a lot of glitches as well on some sounds anyway, but 90% of them were very playable. My setup was using the NI Kontrol driver (which is essentially the same as a normal usb midi driver) and an I2S dac. I don't think those two matter too much though.

I think the main divergence is the fact that I overclocked the Pi a bit, changing the Circle library configuration. I tried changing several configurations in the app but none worked, I ended up changing the Circle library directly locally:

<img width="1352" alt="Image" src="https://github.com/user-attachments/assets/e355052d-df51-4033-b231-ae2f9f2a038d" />

**Comment by michelemaroni** [comment]
> I think the main divergence is the fact that I overclocked the Pi a bit, changing the Circle library configuration. I tried changing several configurations in the app but none worked, I ended up changing the Circle library directly locally:
> 
> <img alt="Image" width="1352" src="https://private-user-images.githubusercontent.com/1353142/432850824-e355052d-df51-4033-b231-ae2f9f2a038d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4MzU5NTMsIm5iZiI6MTc0NTgzNTY1MywicGF0aCI6Ii8xMzUzMTQyLzQzMjg1MDgyNC1lMzU1MDUyZC1kZjUxLTQwMzMtYjIzMS1hZTJmOWYyYTAzOGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjhUMTAyMDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjczMWI3MjhmMzA3ZDViYzRmNDY0NTQwODY1NzE1ZDAxZmNiZjI4ZGZkNGE5MDBjY2JmODk2YzM1OGI4YmZiMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.s418AF43DX3jnYYCORFwWlQFQa6Gu_7Wlf7X4UWPq9A">

Hi @giulioz thanks for sharing this important bit.
Could you check wether you have comparable performance with the default values for Circle's CPUThrottle, but using 
`fast=true`
`socmaxtemp=80`
in cmdline.txt as described in [here](https://github.com/rsta2/circle/blob/master/doc/cmdline.txt)?

I'll test this also on my rpi3 a build and see what happens. @ichingxs4all feel free to do the same on my build as well and report the outcomes. At the moment you are far more advanced in terms of a working and assembled synth, mine is currently dismanteld in a drawer....

**Comment by ichingxs4all** [comment]
Thanks. I will give it a try as soon I'm back from my holidays.
I'm also putting a heatsink tot the Pi Zero 2 to make sure it's not over cooked ;-)

**Comment by GuinanNo1** [comment]
@ichingxs4all Hi, I'm trying to get JV880 working on a BulkyMIDI Dexed hardware build. Runs but no MIDI from the hardware MIDI IN...probably still looking for USB MIDI. Did you make a build for your rackmount that you could share? (also I think I have the same display and more limited button setup as your hardware).

**Comment by sinitsinmike** [comment]
> [@ichingxs4all](https://github.com/ichingxs4all) Hi, I'm trying to get JV880 working on a BulkyMIDI Dexed hardware build. Runs but no MIDI from the hardware MIDI IN...probably still looking for USB MIDI. Did you make a build for your rackmount that you could share? (also I think I have the same display and more limited button setup as your hardware).

For midi all you have to do is set correct/your actual pins in *.ini file, and it will work.

**Comment by ichingxs4all** [comment]
I would recommend to follow the fork of this project:
https://github.com/Sterr1/Mini-JV880pi

Gene did a lot of of working on top of the initial start of giulioz work.
It has serial midi implemented and you can control the buttons of the control interface through midi cc .
That's why I can you use the minimal setup with 2 buttons and a rotary encoder.

Also nice is that the project is near a release of version 1.1 which implements also expansion roms and midi / ftp over network.(Like mt32pi )

I used the euro rack design from DIY ElectroMusic
https://diyelectromusic.com/2025/02/22/minidexed-eurorack-pcb-design/

I recommend in the case of a Raspberry Pi Zero 2W to use a cooling block. And what I also recommend is not to use a regular 7805 but a step-down buck convertor. I used these :
https://www.otronic.nl/nl/step-down-buck-converter-van-45v-24v-naar-5v-3a-4r.html

<img width="1565" height="946" alt="Image" src="https://github.com/user-attachments/assets/d720da9b-bbae-49a8-b083-3f5fb14503c3" />

**Comment by ichingxs4all** [comment]
This is the config file I use for my hardware setup (Eurorack MiniDexed)

[minijv880.ini.zip](https://github.com/user-attachments/files/25737269/minijv880.ini.zip)

**Comment by GuinanNo1** [comment]
Fantastic….thanks for the advice.
This should help me move forward.
I’ll check them out tonight.
Regards, Ian

From: Michel Gutlich ***@***.***>
Reply to: giulioz/mini-jv880 ***@***.***>
Date: Wednesday, 4 March 2026 at 8:36 pm
To: giulioz/mini-jv880 ***@***.***>
Cc: GuinanNo1 ***@***.***>, Comment ***@***.***>
Subject: Re: [giulioz/mini-jv880] Build fails (Issue #1)

[Image removed by sender.]ichingxs4all left a comment (giulioz/mini-jv880#1)<https://github.com/giulioz/mini-jv880/issues/1#issuecomment-3996347258>

I would recommend to follow the fork of this project:
https://github.com/Sterr1/Mini-JV880pi

Gene did a lot of of working on top of the initial start of giulioz work.
It has serial midi implemented and you can control the buttons of the control interface through midi cc .
That's why I can you use the minimal setup with 2 buttons and a rotary encoder.

Also nice is that the project is near a release of version 1.1 which implements also expansion roms and midi / ftp over network.(Like mt32pi )

I used the euro rack design from DIY ElectroMusic
https://diyelectromusic.com/2025/02/22/minidexed-eurorack-pcb-design/

I recommend in the case of a Raspberry Pi Zero 2W to use a cooling block. And what I also recommend is not to use a regular 7805 but a step-down buck convertor. I used these :
https://www.otronic.nl/nl/step-down-buck-converter-van-45v-24v-naar-5v-3a-4r.html
EuroRack-MiniDexed-mt32-jv880-TX816.png (view on web)<https://github.com/user-attachments/assets/d720da9b-bbae-49a8-b083-3f5fb14503c3>
